src/Microsoft.Health.Fhir.Shared.Core/Features/Operations/MemberMatch/MemberMatchService.cs

```diff
@@ -0,0 +1,142 @@
+            _resourceDeserializer = resourceDeserializer;
+            _searchIndexer = searchIndexer;
+            _expressionParser = expressionParser;
+            _coverageBeneficiaryParameter = searchParameterDefinitionManagerResolver().GetSearchParameter("Coverage", "beneficiary");
+        }
+
+        public async Task<ResourceElement> FindMatch(ResourceElement coverage, ResourceElement patient, CancellationToken cancellationToken)
+        {
+            var coverageValues = _searchIndexer.Extract(coverage);
+            var patientValues = _searchIndexer.Extract(patient);
+            var expressions = new List<Expression>();
+            var reverseChainExpressions = new List<Expression>();
+            var searchOptions = new SearchOptions();
+            searchOptions.MaxItemCount = 2;
+            searchOptions.Sort = new List<(SearchParameterInfo, SortOrder)>();
+            searchOptions.UnsupportedSearchParams = new List<Tuple<string, string>>();
+
+            foreach (var patientValue in patientValues)
+            {
+                if (IgnoreInSearch(patientValue))
+                {
+                    continue;
+                }
+
+                var modifier = string.Empty;
+                if (patientValue.SearchParameter.Type == ValueSets.SearchParamType.String)
+                {
+                    modifier = ":exact";
+                }
+
+                expressions.Add(_expressionParser.Parse(new[] { "Patient" }, patientValue.SearchParameter.Code + modifier, patientValue.Value.ToString()));
+            }
+
+            foreach (var coverageValue in coverageValues)
+            {
+                if (IgnoreInSearch(coverageValue))
+                {
+                    continue;
+                }
+
+                var modifier = string.Empty;
+                if (coverageValue.SearchParameter.Type == ValueSets.SearchParamType.String)
+                {
+                    modifier = ":exact";
+                }
+
+                reverseChainExpressions.Add(_expressionParser.Parse(new[] { "Coverage" }, coverageValue.SearchParameter.Code + modifier, coverageValue.Value.ToString()));
+            }
+
+            Expression reverseChainedExpression;
+            if (reverseChainExpressions.Count == 1)
+            {
+                reverseChainedExpression = reverseChainExpressions[0];
+            }
+            else
+            {
+                reverseChainedExpression = Expression.And(reverseChainExpressions);
+            }
+
+            var expression = Expression.Chained(new string[] { "Coverage" }, _coverageBeneficiaryParameter, new string[] { "Patient" }, true, reverseChainedExpression);
+            expressions.Add(expression);
+            searchOptions.Expression = Expression.And(expressions);
+            using IScoped<ISearchService> search = _searchServiceFactory();
+
+            SearchResult results = await search.Value.SearchAsync(searchOptions, cancellationToken);
```

Comment:
>SearchAsync [](http://example.com/codeflow?start=54&length=11)

need to wrap it in try catch (Cosmos founds more than 100 chain resource, sql can't create plan etc)
also need to add logging to log all other exceptions and return 500.